File: backend/app/watchlist/schemas.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.data_sources.domain import AssetType, MarketStatus
from app.overview.schemas import DataSourceSummary


class WatchlistItemCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    thscode: str = Field(min_length=4, max_length=20)
    note: str | None = Field(default=None, max_length=1000)

    @field_validator("thscode")
    @classmethod
    def normalize_thscode(cls, value: str) -> str:
        return value.upper()

    @field_validator("note")
    @classmethod
    def normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class WatchlistFromRadarCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    thscode: str = Field(min_length=9, max_length=9)
    name: str = Field(min_length=1, max_length=120)
    industry: str | None = Field(default=None, max_length=100)

    @field_validator("thscode")
    @classmethod
    def normalize_radar_thscode(cls, value: str) -> str:
        normalized = value.upper()
        if (
            len(normalized) != 9
            or not normalized[:6].isdigit()
            or normalized[6:] not in {".SH", ".SZ"}
        ):
            raise ValueError("红利雷达只支持沪深 A 股")
        return normalized

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("股票名称不能为空")
        return normalized

    @field_validator("industry")
    @classmethod
    def normalize_industry(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class WatchlistItemUpdate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    note: str | None = Field(default=None, max_length=1000)

    @field_validator("note")
    @classmethod
    def normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None

    @model_validator(mode="after")
    def require_submitted_field(self) -> "WatchlistItemUpdate":
        if not self.model_fields_set:
            raise ValueError("请至少提交一个需要修改的字段")
        return self


class WatchlistOrderPayload(BaseModel):
    item_ids: list[uuid.UUID] = Field(min_length=1)


class WatchlistBatchDeletePayload(BaseModel):
    item_ids: list[uuid.UUID] = Field(min_length=1)


class WatchlistItemResponse(BaseModel):
    id: uuid.UUID
    thscode: str
    ticker: str
    name: str
    asset_type: AssetType
    exchange: str
    industry: str | None
    note: str | None
    sort_order: int
    added_at: datetime
    created_at: datetime
    updated_at: datetime
    latest: float | None = None
    change: float | None = None
    change_percent: float | None = None
    volume: float | None = None
    turnover: float | None = None
    total_market_cap: float | None = None
    pe_ttm: float | None = None
    pe_dynamic: float | None = None
    pb: float | None = None
    dividend_yield: float | None = None
    concept: str | None = None
    volume_ratio: float | None = None
    turnover_rate: float | None = None
    quoted_at: datetime | None = None


class WatchlistListResponse(BaseModel):
    items: list[WatchlistItemResponse]
    data_source: DataSourceSummary
    market_status: MarketStatus
    polling_enabled: bool
    refresh_seconds: int
    stale: bool = False


class MessageResponse(BaseModel):
    message: str

File: .pnpm-config/pnpm/store/v11/projects/d2bd198cdd43630583dbbe9675e8d12c/src/features/watchlist/schemas.ts
import { z } from "zod";

const assetTypeSchema = z.enum(["a_share", "fund_etf"]);
const marketStatusSchema = z.enum(["交易中", "午间休市", "已收盘", "休市", "未知"]);

const watchlistItemSchema = z
  .object({
    id: z.string(),
    thscode: z.string(),
    ticker: z.string(),
    name: z.string(),
    asset_type: assetTypeSchema,
    exchange: z.string(),
    industry: z.string().nullable(),
    note: z.string().nullable(),
    sort_order: z.number(),
    added_at: z.string(),
    created_at: z.string(),
    updated_at: z.string(),
    latest: z.number().nullable(),
    change: z.number().nullable(),
    change_percent: z.number().nullable(),
    volume: z.number().nullable(),
    turnover: z.number().nullable(),
    total_market_cap: z.number().nullable(),
    pe_ttm: z.number().nullable(),
    pe_dynamic: z.number().nullable(),
    pb: z.number().nullable(),
    dividend_yield: z.number().nullable(),
    concept: z.string().nullable(),
    volume_ratio: z.number().nullable(),
    turnover_rate: z.number().nullable(),
    quoted_at: z.string().nullable(),
  })
  .passthrough();

export const watchlistResponseSchema = z
  .object({
    items: z.array(watchlistItemSchema),
    data_source: z
      .object({
        state: z.enum(["ready", "not_configured", "authentication_failed", "rate_limited", "unavailable"]),
        name: z.string().nullable(),
        message: z.string().nullable(),
      })
      .passthrough(),
    market_status: marketStatusSchema,
    polling_enabled: z.boolean(),
    refresh_seconds: z.number(),
    stale: z.boolean(),
  })
  .passthrough();
